### PR TITLE
SQSCANNER-23 Support the new 'MessageException' unchecked exception a…

### PR DESCRIPTION
**Pull request by dbmeneses**
…nd log by default the error stack trace only when a non-MessageException is thrown

### CODE REVIEW

File: src/main/java/org/sonarsource/scanner/cli/Main.java [hunk 1]
```diff
@@ -147,13 +135,18 @@ private void showError(String message, Throwable e, boolean showStackTrace) {
...
       logger.error("");
-      logger.error("To see the full stack trace of the errors, re-run SonarQube Scanner with the -e switch.");
-      if (!cli.isDebugEnabled()) {
-        suggestDebugMode();
-      }
+      suggestDebugMode();
     }
   }
+  
+  private boolean showStackTrace(Throwable e, boolean debug) {
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/cli/Main.java]: MAJOR Make "showStackTrace" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: src/main/java/org/sonarsource/scanner/cli/Main.java [hunk 2]
```diff
@@ -147,13 +135,18 @@ private void showError(String message, Throwable e, boolean showStackTrace) {
...
+    // class not available at compile time (loaded by isolated classloader)
+    return debug || "org.sonar.api.utils.MessageException".equals(e.getClass().getName());
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/cli/Main.java]: MAJOR Use an "instanceof" comparison instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1872)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: src/main/java/org/sonarsource/scanner/cli/Main.java [hunk 3]
```diff
@@ -147,13 +135,18 @@ private void showError(String message, Throwable e, boolean showStackTrace) {
...
-      if (!cli.isDebugEnabled()) {
-        suggestDebugMode();
-      }
+      suggestDebugMode();
     }
   }
+  
+  private static boolean showStackTrace(Throwable e, boolean debug) {
+    // class not available at compile time (loaded by isolated classloader)
+    return debug || "org.sonar.api.utils.MessageException".equals(e.getClass().getName());
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/cli/Main.java]: MAJOR Use an "instanceof" comparison instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1872)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: src/main/java/org/sonarsource/scanner/cli/Cli.java [hunk 4]
```diff
@@ -66,10 +61,9 @@ Cli parse(String[] args) {
 
       } else if ("-v".equals(arg) || "--version".equals(arg)) {
         displayVersionOnly = true;
-
+
```
Comment (janos-ss) [inline src/main/java/org/sonarsource/scanner/cli/Cli.java]: I saw in your other PRs removal of such trailing spaces (good)... Now they seem to be creeping back. Issue with Eclipse formatting settings?

File: src/test/java/org/sonarsource/scanner/cli/MainTest.java [hunk 5]
```diff
@@ -95,33 +96,36 @@ public void should_stop_on_error() {
...
-  public void show_error_stacktrace() {
-    Exception e = show_error(false, true);
-    verify(logs).error("Error during SonarQube Scanner execution", e);
+  public void show_error() {
+    show_error(false, false);
+    verify(logs).error("Error during SonarQube Scanner execution");
     verify(logs).error("Re-run SonarQube Scanner using the -X switch to enable full debug logging.");
   }
-
+
```
Comment (janos-ss) [inline src/test/java/org/sonarsource/scanner/cli/MainTest.java]: another example of issue with Eclipse formatting settings...

File: src/test/java/org/sonarsource/scanner/cli/MainTest.java [hunk 6]
```diff
@@ -95,33 +96,36 @@ public void should_stop_on_error() {
...
+    Exception e = show_error(true, false);
 
     verify(logs).error("Error during SonarQube Scanner execution", e);
     verify(logs, never()).error("Re-run SonarQube Scanner using the -X switch to enable full debug logging.");
   }
 
-  private Exception show_error(boolean debugEnabled, boolean stackTraceEnabled) {
-    Exception e = new NullPointerException("NPE");
-    e = new IllegalStateException("Error", e);
+  private Exception show_error(boolean debugEnabled, boolean messageException) {
```
Comment (janos-ss) [inline src/test/java/org/sonarsource/scanner/cli/MainTest.java]: This method *smells*... The name suggests to "show" something, at the same time it also creates and returns an `Exception`. Also, `snake_case` is fine in test methods, but this is not a test method.

File: src/main/java/org/sonarsource/scanner/cli/Cli.java (hunk 4 above)
Comment (dbmeneses) [inline src/main/java/org/sonarsource/scanner/cli/Cli.java]: yes, fixed

File: src/test/java/org/sonarsource/scanner/cli/MainTest.java (hunk 6 above)
Comment (dbmeneses) [inline src/test/java/org/sonarsource/scanner/cli/MainTest.java]: fixed

**Comment by ddwolf** [comment]
image
Avg Complexity(Statements) and Avg Complexity(Methods), What are they?Sombe body help explaining these?

**Pull request merged**